script.module.cache/lib/cache.py

```diff
@@ -0,0 +1,307 @@
+# -*- coding: utf-8 -*-
+
+"""HTTP File cache and helpers"""
+
+__author__ = "fraser"
+
+import logging
+import sqlite3
+import time
+from datetime import datetime, timedelta, tzinfo
+from os import path
+
+import xbmc
+import xbmcaddon
+import xbmcvfs
+
+try:
+    import _pickle as cpickle
+except ImportError:
+    import cPickle as cpickle
+
+logger = logging.getLogger(__name__)
+
+HTTP_DATE_FORMAT = "%a, %d %b %Y %H:%M:%S %Z"
+ADDON = xbmcaddon.Addon()
+ADDON_PROFILE = ADDON.getAddonInfo("profile").encode("utf-8").decode("utf-8")
+CACHE_URI = xbmc.translatePath(path.join(ADDON_PROFILE, "cache.sqlite"))
+
+
+def httpdate_to_datetime(input_date):
+    # type: (str) -> Optional[datetime]
+    """Converts http date string to datetime"""
+    if input_date is None:
+        return None
+    try:
+        return datetime.strptime(input_date, HTTP_DATE_FORMAT)
+    except TypeError:
+        return datetime(*(time.strptime(input_date, HTTP_DATE_FORMAT)[0:6]))
+    except ValueError as e:
+        logger.debug(e)
+        return None
+
+
+def datetime_to_httpdate(input_date):
+    # type: (datetime) -> Optional[datetime]
+    """Converts datetime to http date string"""
+    if input_date is None:
+        return None
+    try:
+        return input_date.strftime(HTTP_DATE_FORMAT)
+    except (ValueError, TypeError) as e:
+        logger.debug(e)
+        return None
+
+
+def conditional_headers(row):
+    # type: (sqlite3.Row) -> dict
+    """Creates dict of If-None-Match and If-Modified-Since headers based on row etag and last_modified columns"""
+    headers = {}
+    if row["etag"] is not None:
+        headers["If-None-Match"] = row["etag"]
+    if row["last_modified"] is not None:
+        headers["If-Modified-Since"] = datetime_to_httpdate(row["last_modified"])
+    return headers
+
+
+class Store(object):
+    """
+    Generic unique string storage helper
+    Saves unique strings in a set that can
+    be retrieved, appended to, removed from or cleared entirely
+    """
+
+    def __init__(self, key, db=None):
+        # type: (str, str) -> None
+        """
+        Creates a new Store object
+        key is the identifier that the set of strings are stored under
+        db is the path to a sqlite3 database (defaults to Cache default)
+        """
+        self.db = db
+        self.key = key
+
+    def retrieve(self):
+        # type: () -> list
+        """Gets set of stored strings"""
+        with Cache(self.db) as c:
+            data = c.get(self.key)
+            return data["blob"] if data else set()
+
+    def _save(self, data):
+        # type: (list) -> None
+        """Saves set of strings"""
+        with Cache(self.db) as c:
+            if type(data) is set:
+                c.set(self.key, data)
+
+    def append(self, item):
+        # type: (str) -> None
+        """Add string to the store"""
+        current = self.retrieve()
+        current.add(item)
```

Comment:
Another Python 3 incompatibility added in yesterdays changes.
```
  File "<...>\addons\script.module.cache\lib\cache.py", line 102, in append
    current.add(item)
AttributeError: 'str' object has no attribute 'add'
```

Reply:
Pretty sure that is because of a cached string value, "current' should be of type Set...I think clearing the cache would fix this. In any case I have amended the retrieve method to ensure the value is a set. Done